File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/cpp/au_015_c_add_u64_f64.cpp
// arch-unit 15  --  c  `a + b`  lhs=uint64_t rhs=double
// symbol op_118   outcome LIFTED
//
// the arch-unit, arch-opcode by arch-opcode:
//   fcvt.d.lu fa5, a0, dyn             float    emulation:ui64_to_f64_rm0
//   fadd.d fa0, fa0, fa5, dyn          float    emulation:f64_add_rm0
//   c.jr ra                            integer  return
//
// answer: f64, 64 bits.  parameters are operand bit patterns.
#include "arch_units.hpp"

namespace archunits {

uint64_t au_015_c_add_u64_f64(uint64_t p0, uint64_t p1)
{
    /* a0: operand `a` (uint64_t) arrives in a0 */
    const uint64_t v1 = p0;
    /* fa0: operand `b` (double) arrives in fa0 */
    const uint64_t v2 = p1;
    const uint64_t v3 = sfemul::ui64_to_f64_rm0(v1);
    const uint64_t v4 = sfemul::f64_add_rm0(v2, v3);
    return v4;
}

}  // namespace archunits

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/rust/au_015_c_add_u64_f64.rs
// arch-unit 15  --  c  `a + b`  lhs=uint64_t rhs=double
// symbol op_118   outcome LIFTED
//
// the arch-unit, arch-opcode by arch-opcode:
//   fcvt.d.lu fa5, a0, dyn             float    emulation:ui64_to_f64_rm0
//   fadd.d fa0, fa0, fa5, dyn          float    emulation:f64_add_rm0
//   c.jr ra                            integer  return
//
// answer: f64, 64 bits.  parameters are operand bit patterns.
#![allow(unused_parens, unused_variables, unused_imports, clippy::all)]
use crate::b2u;

pub fn au_015_c_add_u64_f64(p0: u64, p1: u64) -> u64 {
    // a0: operand `a` (uint64_t) arrives in a0
    let v1: u64 = p0;
    // fa0: operand `b` (double) arrives in fa0
    let v2: u64 = p1;
    let v3: u64 = sfemul::ui64_to_f64::ui64_to_f64_rm0(v1);
    let v4: u64 = sfemul::f64_add::f64_add_rm0(v2, v3);
    v4
}
